File: windows/VoxFlow/Log.cs
using System;
using System.IO;
using System.Text;

namespace VoxFlow;

/// <summary>
/// Append-only diagnostic log at %APPDATA%\VoxFlow\voxflow.log.
/// Deliberately dependency-free and exception-proof: logging must never be
/// the reason the app fails.
/// </summary>
public static class Log
{
    private static readonly object Gate = new();
    private static string? _path;

    public static string Path
    {
        get
        {
            if (_path == null)
            {
                string dir = System.IO.Path.Combine(
                    Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "VoxFlow");
                try { Directory.CreateDirectory(dir); } catch { }
                _path = System.IO.Path.Combine(dir, "voxflow.log");
            }
            return _path;
        }
    }

    public static void Info(string message) => Write("INFO ", message);
    public static void Warn(string message) => Write("WARN ", message);

    public static void Error(string message, Exception? ex = null)
    {
        var sb = new StringBuilder(message);
        if (ex != null)
        {
            sb.Append(Environment.NewLine).Append(ex.ToString());
            var inner = ex.InnerException;
            int depth = 0;
            while (inner != null && depth++ < 5)
            {
                sb.Append(Environment.NewLine).Append("--- inner ---").Append(Environment.NewLine).Append(inner.ToString());
                inner = inner.InnerException;
            }
        }
        Write("ERROR", sb.ToString());
    }

    private static void Write(string level, string message)
    {
        try
        {
            lock (Gate)
            {
                // Keep the log from growing without bound across a long-running install.
                try
                {
                    var fi = new FileInfo(Path);
                    if (fi.Exists && fi.Length > 1_000_000)
                        File.WriteAllText(Path, $"{Stamp()} INFO  (log truncated){Environment.NewLine}");
                }
                catch { }

                File.AppendAllText(Path, $"{Stamp()} {level} {message}{Environment.NewLine}");
            }
        }
        catch { /* logging must never throw */ }
    }

    private static string Stamp() => DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss.fff");
}

File: macos/Sources/VoxFlow/Support/Log.swift
import Foundation
import os

/// Append-only diagnostic log at
/// `~/Library/Application Support/VoxFlow/voxflow.log`, mirrored to the
/// unified log. Deliberately dependency-free and exception-proof: logging
/// must never be the reason the app fails.
///
/// The unified log is fine for developers with Console.app open, but a user
/// whose hotkey has just died needs something they can open from the menu
/// bar and paste into a bug report — hence the file (same as the Windows
/// build's `%APPDATA%\VoxFlow\voxflow.log`).
enum Log {
    private static let gate = NSLock()
    private static let unified = Logger(subsystem: "com.voxflow.app", category: "VoxFlow")
    private static let maxBytes: UInt64 = 1_000_000

    static let fileURL: URL = {
        let dir = PersonalDictionary.applicationSupportDirectory()
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir.appendingPathComponent("voxflow.log")
    }()

    static func info(_ message: String) {
        unified.info("\(message, privacy: .public)")
        write("INFO ", message)
    }

    static func warn(_ message: String) {
        unified.notice("\(message, privacy: .public)")
        write("WARN ", message)
    }

    static func error(_ message: String, _ error: Error? = nil) {
        var text = message
        if let error {
            text += "\n" + String(describing: error)
            let ns = error as NSError
            if let underlying = ns.userInfo[NSUnderlyingErrorKey] as? NSError {
                text += "\n--- underlying ---\n" + String(describing: underlying)
            }
        }
        unified.error("\(text, privacy: .public)")
        write("ERROR", text)
    }

    private static func write(_ level: String, _ message: String) {
        gate.lock()
        defer { gate.unlock() }
        let line = "\(stamp()) \(level) \(message)\n"
        guard let data = line.data(using: .utf8) else { return }
        do {
            let fm = FileManager.default
            if let attrs = try? fm.attributesOfItem(atPath: fileURL.path),
               let size = attrs[.size] as? UInt64, size > maxBytes {
                // Keep the log from growing without bound across a long-running install.
                try? "\(stamp()) INFO  (log truncated)\n".data(using: .utf8)?.write(to: fileURL)
            }
            if !fm.fileExists(atPath: fileURL.path) {
                try data.write(to: fileURL)
                return
            }
            let handle = try FileHandle(forWritingTo: fileURL)
            defer { try? handle.close() }
            try handle.seekToEnd()
            try handle.write(contentsOf: data)
        } catch {
            // logging must never throw
        }
    }

    private static let stampFormatter: DateFormatter = {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.dateFormat = "yyyy-MM-dd HH:mm:ss.SSS"
        return f
    }()

    private static func stamp() -> String { stampFormatter.string(from: Date()) }

    /// Milliseconds elapsed since a reference `Date`.
    static func ms(since start: Date) -> Int {
        Int(Date().timeIntervalSince(start) * 1000)
    }
}
